src/managed/Microsoft.NET.HostModel/AppHost/HostWriter.cs

```diff
@@ -43,45 +44,68 @@ public static class HostWriter
+                    {
+                        BinaryUtils.SearchAndReplace(accessor, AppBinaryPathPlaceholderSearchValue, bytesToWrite);
 
-                    appHostIsPEImage = BinaryUtils.IsPEImage(accessor);
+                        appHostIsPEImage = BinaryUtils.IsPEImage(accessor);
 
-                    if (windowsGraphicalUserInterface)
-                    {
-                        if (!appHostIsPEImage)
+                        if (windowsGraphicalUserInterface)
                         {
-                            throw new AppHostNotPEFileException();
+                            if (!appHostIsPEImage)
+                            {
+                                throw new AppHostNotPEFileException();
+                            }
+
+                            BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
                         }
+                    }
+                }
 
-                        BinaryUtils.SetWindowsGraphicalUserInterfaceBit(accessor);
+                if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                {
+                    if (ResourceUpdater.IsSupportedOS())
+                    {
+                        // Copy resources from managed dll to the apphost
+                        new ResourceUpdater(appHostDestinationFilePath)
+                            .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
+                            .Update();
+                    }
+                    else
+                    {
+                        throw new AppHostCustomizationUnsupportedOSException();
                     }
                 }
-            }
 
-            if (assemblyToCopyResorcesFrom != null && appHostIsPEImage)
+                // Memory-mapped write does not updating last write time
+                File.SetLastWriteTimeUtc(appHostDestinationFilePath, DateTime.UtcNow);
+            }
+            catch (Exception ex)
             {
-                if (ResourceUpdater.IsSupportedOS())
+                FailedToDeleteApphostException failedToDeleteApphostException = null;
+                // Delete the destination file so we don't leave an unmodified apphost
+                try
                 {
-                    // Copy resources from managed dll to the apphost
-                    new ResourceUpdater(appHostDestinationFilePath)
-                        .AddResourcesFromPEImage(assemblyToCopyResorcesFrom)
-                        .Update();
+                    File.Delete(appHostDestinationFilePath);
                 }
-                else 
+                catch (Exception failedToDeleteEx) when (failedToDeleteEx is IOException || failedToDeleteEx is UnauthorizedAccessException)
                 {
-                    throw new AppHostCustomizationUnsupportedOSException();
+                    failedToDeleteApphostException = new FailedToDeleteApphostException(failedToDeleteEx.Message);
```

Comment:
I would put failedToDeleteEx in innner exception so that the AggregateException has all stack traces.

Reply:
Actually, do we need FailedToDeleteException at all? Can we just throw an AggregateException of the deletion exception and the outer exception? 

The goal was not to lose the info if Delete is also failing, but an arbitrary exception here is going to become a big scary stack trace in MSBuild, right? If we aggregate both, we'll have all the info in that bug report. The SDK wouldn't need to handle AggregateException specially and log it.

Reply:
I would also remove the when from the Delete catch.

``` C#
try
{
    // create apphost
}
catch (Exception ex)
{
     try
     {
            // delete apphost
     }
     catch (Exception deleteEx)
     {
            throw new AggregateException(ex, deleteEx);
     }

     throw;
}
```